src/services/classifier2020.ts

```diff
@@ -0,0 +1,257 @@
+/** @internal */
+namespace ts.classifier.modern {
+
+    /** @internal */
+    export enum TokenEncodingConsts {
+        typeOffset = 8,
+        modifierMask = (1 << typeOffset) - 1
+    }
+
+    /** @internal */
+    export enum TokenType {
+        class, enum, interface, namespace, typeParameter, type, parameter, variable, enumMember, property, function, member, _
+    }
+
+    /** @internal */
+    export enum TokenModifier {
+        declaration, static, async, readonly, defaultLibrary, local, _
+    }
+
+    /** This is mainly used internally for testing */
+    export function getSemanticClassifications(program: Program, _cancellationToken: CancellationToken, sourceFile: SourceFile, span: TextSpan): ClassifiedSpan[] {
+        const classifications = getEncodedSemanticClassifications(program, _cancellationToken, sourceFile, span);
+
+        Debug.assert(classifications.spans.length % 3 === 0);
+        const dense = classifications.spans;
+        const result: ClassifiedSpan[] = [];
+        for (let i = 0; i < dense.length; i += 3) {
+                result.push({
+                        textSpan: createTextSpan(dense[i], dense[i + 1]),
+                        classificationType: dense[i + 2]
+                });
+        }
+
+        return result;
+    }
+
+    export function getEncodedSemanticClassifications(program: Program, _cancellationToken: CancellationToken, sourceFile: SourceFile, span: TextSpan): Classifications {
+        return {
+            spans: getSemanticTokens(program, sourceFile, span),
+            endOfLineState: EndOfLineState.None
+        };
+    }
+
+    function getSemanticTokens(program: Program, sourceFile: SourceFile, span: TextSpan): number[] {
+        const resultTokens: number[] = [];
+
+        const collector = (node: Node, typeIdx: number, modifierSet: number) => {
+            resultTokens.push(node.getStart(), node.getWidth(), ((typeIdx + 1) << TokenEncodingConsts.typeOffset) + modifierSet);

```

Comment:
You can make this a little speedier if you pass in `sourceFile` to `getStart` and `getWidth`.